### ISSUE TITLE
In tflite, how to use the same memory to serve different models with exactly the same structure

### ISSUE THREAD

**Issue by captainst** [tensorflow/tensorflow#69823]
Hello Everyone,

We have an image classification application where we built different squeezenet models with the exact same model structure.

In the app, we use the model one at a time. In order to save memory and avoid memory fragmentation, we plan to completely reuse the memory, both for storing the static model weights, and for storing the dynamic intermediary tensors. We loaded the 1st model, created the interpreter as usual:

```
_model = tflite::FlatBufferModel::BuildFromFile(model_filename.c_str());
tflite::InterpreterBuilder(*_model, resolver)(&_interpreter);
_interpreter->AllocateTensors();
```

One way to achieve the memory reuse is to modify the model weights on-the-fly. However, the problem is that after the model is built, all the tensors holding weights are of type `TfLiteAllocationType::kTfLiteMmapRo`. I could read the weights correctly. But modifying the data of these tensors always result in a segment fault: 

```
auto tsr = _interpreter->tensor(index);
std::cout << (float)(*tsr->data.data) << std::endl; // this works
*tar->data.data = 0.01f; // this always causes segment fault
```

Does anyone know if there are workarounds to achieve this ?

Many thanks !

**Comment by sushreebarsa** [comment]
@captainst TensorFlow Lite (TFLite) doesn't have a built-in mechanism to directly share memory between different models loaded in separate interpreters. Model quantization is the recommended approach for this issue, kindly refer to [this](https://www.tensorflow.org/lite/performance/post_training_quantization) guide.
Thank you!

**Comment by captainst** [comment]
> @captainst TensorFlow Lite (TFLite) doesn't have a built-in mechanism to directly share memory between different models loaded in separate interpreters. Model quantization is the recommended approach for this issue, kindly refer to [this](https://www.tensorflow.org/lite/performance/post_training_quantization) guide. Thank you!

Thank you! I only need to load one model at a time. So I don't need separate interpreters. Since all the models are of the same structure, I want to use one interpreter to serve all the models, one at a time. If I could find a way to manipulate the weights and quant params inside each tensor.

I'll do some further experiments to see. Thank you again.

**Comment by sushreebarsa** [comment]
@captainst Is there any update on this issue?
Thank you!

**Comment by captainst** [comment]
Hi @sushreebarsa 

I made some investigations through the **subgraph** and **arena_planner**, and found that tflite does not provide no an such interface for the moment. I managed to use `interpreter->SetCustomAllocationForTensor` to mount an external allocated memory to each tensor. This is a technical viable approach, but in this case we loose the efficiency of memory usage provided by arena_planner. The amount memory we needed to provide externally is 2x to 4x, which does not make practical sense.

A more intelligent solution could be to make arena_planner use a memory pool provided externally. This could help to greatly alleviate the memory fragmentation problem, when we need to load and off-load different models frequently. This involves some modifications to arena. 

For the moment I would think that the issue being closed. Probably we could contribute with PR when it is done from our side.
